File: src/apps/archspace/turn_processor.cc
/**
 * Simple turn processor helper
 * This ensures turns are processed even if the main update thread has issues
 */

#include <cstdio>
#include <ctime>
#include <unistd.h>
#include <mysql/mysql.h>

int main() {
    printf("Turn Processor Helper Starting...\n");
    
    MYSQL *conn = mysql_init(NULL);
    if (conn == NULL) {
        fprintf(stderr, "mysql_init() failed\n");
        return 1;
    }
    
    // Get database credentials from environment
    const char* host = getenv("DB_HOST") ? getenv("DB_HOST") : "mysql";
    const char* user = getenv("DB_USER") ? getenv("DB_USER") : "archspace";
    const char* pass = getenv("DB_PASSWORD") ? getenv("DB_PASSWORD") : "archspace123";
    const char* db = getenv("DB_NAME") ? getenv("DB_NAME") : "Archspace2";
    
    if (mysql_real_connect(conn, host, user, pass, db, 3306, NULL, 0) == NULL) {
        fprintf(stderr, "mysql_real_connect() failed: %s\n", mysql_error(conn));
        mysql_close(conn);
        return 1;
    }
    
    printf("Connected to database successfully\n");
    
    int turn = 0;
    time_t last_turn_time = time(NULL);
    const int turn_interval = 300; // 5 minutes
    
    while (1) {
        time_t now = time(NULL);
        
        if (now - last_turn_time >= turn_interval) {
            turn++;
            printf("[%ld] Processing turn %d\n", now, turn);
            
            // Update all players' turns
            char query[512];
            snprintf(query, sizeof(query), 
                "UPDATE player SET turn = %d, tick = %ld WHERE game_id > 0 AND game_id != 9999999",
                turn, now + turn_interval);
            
            if (mysql_query(conn, query)) {
                fprintf(stderr, "UPDATE failed: %s\n", mysql_error(conn));
            } else {
                printf("Updated %llu players\n", mysql_affected_rows(conn));
            }
            
            // Simple resource update
            snprintf(query, sizeof(query),
                "UPDATE player SET production = production + 100, research = research + 10 WHERE game_id > 0 AND game_id != 9999999");
            mysql_query(conn, query);
            
            last_turn_time = now;
        }
        
        sleep(10); // Check every 10 seconds
    }
    
    mysql_close(conn);
    return 0;
}

File: turn_processor.php
#!/usr/bin/env php
<?php
echo "[TURNS] Starting turn processor\n";
echo "[TURNS] PHP Version: " . PHP_VERSION . "\n";

while (true) {
    try {
        // Use Railway's environment variables
        $db_host = getenv('DB_HOST') ?: getenv('MYSQLHOST') ?: 'localhost';
        $db_name = getenv('DB_NAME') ?: getenv('MYSQLDATABASE') ?: 'Archspace2';
        $db_user = getenv('DB_USER') ?: getenv('MYSQLUSER') ?: 'archspace';
        $db_pass = getenv('DB_PASSWORD') ?: getenv('MYSQLPASSWORD') ?: 'archspace123';
        
        echo "[TURNS] Connecting to $db_host/$db_name as $db_user\n";
        
        $pdo = new PDO("mysql:host=$db_host;dbname=$db_name", $db_user, $db_pass);
        $stmt = $pdo->query("SELECT MIN(tick) as next_tick, MAX(turn) as current_turn FROM player WHERE game_id > 0 AND game_id != 9999999");
        $info = $stmt->fetch(PDO::FETCH_ASSOC);
        
        $currentTime = time();
        $nextTick = $info['next_tick'] ?: 0;
        $currentTurn = $info['current_turn'] ?: 0;
        
        echo "[TURNS] Current time: $currentTime, Next tick: $nextTick, Current turn: $currentTurn\n";
        
        // Process if tick is in the past or invalid
        if ($nextTick <= $currentTime || $nextTick == 0 || ($currentTime - $nextTick) > 600) {
            $newTurn = $currentTurn + 1;
            $newTick = $currentTime + 300;
            
            echo "[TURNS] Processing turn $currentTurn -> $newTurn\n";
            
            $stmt = $pdo->prepare("UPDATE player SET turn = ?, tick = ? WHERE game_id > 0 AND game_id != 9999999");
            $stmt->execute([$newTurn, $newTick]);
            
            $pdo->exec("UPDATE player SET production = production + 100, research = research + 10, military = military + 5 WHERE game_id > 0 AND game_id != 9999999");
            
            echo "[TURNS] Turn $newTurn complete!\n";
        }
    } catch (Exception $e) {
        echo "[TURNS] Error: " . $e->getMessage() . "\n";
        echo "[TURNS] Stack trace: " . $e->getTraceAsString() . "\n";
    }
    sleep(10); // Check every 10 seconds
}
?>